docs/design-docs/08-package-variant.md

```diff
@@ -0,0 +1,234 @@
+made to the upstream specification in the resource, and the controller would
+propose a new Draft package reflecting the outcome of `kpt pkg update`.
+
+By automating this process, we open up the possibility of performing systematic
+changes that tie back to our different dimensions of scalability. For example,
+we can use data about the specific variant we are creating to lookup additional
+context in the Porch cluster, and copy that information into the variant. We
+refer to this as *injection*, and it enables dynamic, context-aware creation of
+variants. This will be explained in more detail below.
+
+This proposal also introduces a way to "fan-out", or create multiple
+`PackageVariant` resources declaratively based upon a list or selector with the
+`PackageVariantSet` resource.. This is combined with the injection mechanism to
+enable generation of large sets of variants that are specialized to a particular
+target repository, cluster, or other target.
+
+## Basic Package Cloning
+
+The `PackageVariant` resource controls the creation and lifecycle of a variant
+of a package. That is, it defines the original (upstream) package, the new
+(downstream) package, and the changes (mutations) that need to be made to
+transform the upstream into the downstream. It also allows the user to specify
+policies around adoption, deletion, and update of package revisions that are
+under the control of the package variant controller.
+
+![Basic Package Variant](PackageVariant%20-%201.png)
+
+Note that *proposal* and *approval* are not handled by the package variant
+controller. Those are left to humans or other controllers. The exception is the
+proposal of deletion (there is no concept of a "Draft" deletion), which the
+package variant control will do, depending upon the specified deletion policy.
+
+### Package Revision Metadata
+
+The package variant controller utilizes Porch APIs. This means that it is not
+just doing a `clone` operation, but in fact creating a Porch `PackageRevision`
+resource. In particular, that resource can contain Kubernetes metadata that is
+not part of the package as stored in the repository.
+
+Some of that metadata is necessary for the management of the PackageRevision by
+the PackageVariant controller - for example, the `ownerRef` indicating which
+`PackageVariant` created the `PackageRevision`. These are not in control of the
+user. However, the `PackageVariant` resource does make the annotations and
+labels of the `PackageRevision` available as values that may be controlled
+during the creation of the `PackageRevision`. This can assist in additional
+automation workflows.
+
+## Introducing Variance
+Just cloning is not that interesting, so the `PackageVariant` resource also
+allows you to control various ways of mutating the original package to create
+the variant.
+
+### Package Context
+[Note: I believe right now we *always* produce `--for-deployment` package
```

Comment:
So the PackageVariant controller doesn't actually do this right now. It just does a clone, without adding or changing any existing `kptfile.kpt.dev` ConfigMap. Whether it _should_ is an interesting idea. It might be useful to have a similar `for-deployment` option in the PackageVariant, since it would keep this controller aligned with the variant constructor pattern that we have with kpt.

Reply:
Ok, good to know. Yeah, I think we probably need a "forDeployment" flag.

Reply:
I think this might be done by Porch, regardless of the `for-deployment` thing. I think the name replacement doesn't happen unless you do `for-deployment`. But otherwise, I think the actual config map gets created automatically already. @mortent do you know?

Reply:
I covered this in the detailed design section for the package context feature.